tests/app/dao/notification_dao/test_notification_dao_template_usage.py

```diff
@@ -0,0 +1,167 @@
+        sample_template):
+    results = dao_get_last_template_usage(sample_template.id, 'sms')
+    assert not results
+
+
+def test_should_by_able_to_get_template_count(sample_template, sample_email_template):
+    create_notification(sample_template)
+    create_notification(sample_template)
+    create_notification(sample_template)
+    create_notification(sample_email_template)
+    create_notification(sample_email_template)
+
+    results = dao_get_template_usage(sample_template.service_id, date.today())
+    assert results[0].name == 'Email Template Name'
+    assert results[0].template_type == 'email'
+    assert results[0].count == 2
+
+    assert results[1].name == 'Template Name'
+    assert results[1].template_type == 'sms'
+    assert results[1].count == 3
+
+
+def test_template_usage_should_ignore_test_keys(
+    sample_team_api_key,
+    sample_test_api_key,
+    sample_api_key,
+    sample_template
+):
+
+    create_notification(sample_template, api_key=sample_api_key, key_type=KEY_TYPE_NORMAL)
+    create_notification(sample_template, api_key=sample_team_api_key, key_type=KEY_TYPE_TEAM)
+    create_notification(sample_template, api_key=sample_test_api_key, key_type=KEY_TYPE_TEST)
+    create_notification(sample_template)
+
+    results = dao_get_template_usage(sample_template.service_id, date.today())
+    assert results[0].name == 'Template Name'
+    assert results[0].template_type == 'sms'
+    assert results[0].count == 3
+
+
+def test_template_usage_should_filter_by_service(notify_db_session):
+    service_1 = create_service(service_name='test1')
+    service_2 = create_service(service_name='test2')
+    service_3 = create_service(service_name='test3')
+
+    template_1 = create_template(service_1)
+    template_2 = create_template(service_2)  # noqa
+    template_3a = create_template(service_3, template_name='a')
+    template_3b = create_template(service_3, template_name='b')  # noqa
+
+    # two for service_1, one for service_3
+    create_notification(template_1)
+    create_notification(template_1)
+
+    create_notification(template_3a)
+
+    res1 = dao_get_template_usage(service_1.id, date.today())
+    res2 = dao_get_template_usage(service_2.id, date.today())
+    res3 = dao_get_template_usage(service_3.id, date.today())
+
+    assert len(res1) == 1
+    assert res1[0].count == 2
+
+    assert len(res2) == 1
+    assert res2[0].count == 0
+
+    assert len(res3) == 2
+    assert res3[0].count == 1
+    assert res3[1].count == 0
+
+
+def test_template_usage_should_by_able_to_get_zero_count_from_notifications_history_if_no_rows(sample_service):
+    results = dao_get_template_usage(sample_service.id, date.today())
+    assert len(results) == 0
+
+
+def test_template_usage_should_by_able_to_get_zero_count_from_notifications_history_if_no_service():
+    results = dao_get_template_usage(str(uuid.uuid4()), date.today())
```

Comment:
Should this actually be an exception? Can this case happen regularly?

Reply:
I don't expect it to ever happen, but not sure it's worth putting extra logic in to raise an exception. i  view this test as more tests-as-documentation than risk management or anything.